duck/call.py

```diff
@@ -0,0 +1,171 @@
+
+        # Determine if the given argument(s) are present in the Call.
+        args_results = []
+        for idx, actual in enumerate(args):
+            if actual == self.args[idx]:
+                args_results.append(True)
+            else:
+                args_results.append(False)
+
+        results.append(all(args_results))
+
+        # If the given keyword arguments exceed the number of actual keyword
+        # arguments, this cannot be true either.
+        if len(kwargs) > len(self.kwargs):
+            return False
+
+        # Determine if the given keyword argument(s) are present in the Call.
+        kwargs_results = []
+        for k, v in kwargs.items():
+            if k in self.kwargs and self.kwargs[k] == v:
+                kwargs_results.append(True)
+            else:
+                kwargs_results.append(False)
+
+        results.append(all(kwargs_results))
+
+        return all(results)
+
+    def has_exact_args(self, *args, **kwargs):
+        """Returns ``True`` if the call was made with exactly the provided
+        arguments and keyword arguments, and ``False`` otherwise.
+
+        If provided, positional arguments must start at ``0`` and be in order.
+
+        Comparisons are performed with ``==``, not ``is``.
+        """
+        # There must be the same number of arguments.
+        if len(args) != len(self.args):
+            return False
+
+        # Arguments must be identical.
+        for idx, actual in enumerate(args):
+            if actual != self.args[idx]:
+                return False
+
+        # There must be the same number of keyword arguments.
+        if len(kwargs) != len(self.kwargs):
+            return False
+
+        # They must be identical.
+        for k, v in kwargs.items():
+            if k not in self.kwargs or self.kwargs[k] != v:
+                return False
+
+        return True
+
+    def assert_has_args(self, *args, **kwargs):
+        """Calls the ``has_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+        """
+        errors = []
+
+        # Determine if the given arguments are valid.
+        if not self.has_args(*args):
+            errors.append('Given args: {} != Valid args: {}'.format(
+                args,
+                self.args,
+            ))
+
+        # Determine if the given keyword arguments are valid.
+        if not self.has_args(**kwargs):
+            errors.append('Given kwargs: {} != Valid kwargs: {}'.format(
+                kwargs,
+                self.kwargs,
+            ))
+
+        if errors:
+            raise AssertionError('\r\n'.join(errors))
+
+        return True
+
+    def assert_has_exact_args(self, *args, **kwargs):
+        """Calls the ``has_exact_args`` method, and raises ``AssertionError``
+        if ``False`` is returned.
+
+        In order to return useful debugging information, it post-processes
+        similar to ``has_exact_args``, but with additional information.
```

Comment:
There are a couple ways to go about this:

1. You could flip your plan on its head, and have `has_args` and `has_exact_args` call their assertion equivalents in a `try/except`:

```python
def has_args(self, *args, **kwargs):
    try:
        return self.assert_has_args(*args, **kwargs)
    except AssertionError:
        return False
```

This is not the most performant approach ever, but for what we are doing (running tests at build time) it is entirely acceptable.

2. You could have `.has_args` return a specialized class explaining the error, with an appropriate `__bool__` method (note that it is spelled `__nonzero__` in Python 2). This object could collect both the arguments that matched and those that did not. I think I actually like this idea the better, if it has a good `__repr__`, because it will make the py.test output amazing.

3. You could have a private argument sufficiently unlikely to be used, such as `_duck_verbose`.